examples/Python/PET/interactive/spdhgutils.py

```diff
@@ -0,0 +1,203 @@
+# gradient algorithm (SPDHG) for regularized PET reconstruction.
+#
+#    [CERS2018] A. Chambolle, M. J. Ehrhardt, P. Richtarik and C.-B. Schoenlieb,
+#    *Stochastic Primal-Dual Hybrid Gradient Algorithm with Arbitrary Sampling
+#    and Imaging Applications*. SIAM Journal on Optimization, 28(4), 2783–2808
+#    (2018) http://doi.org/10.1007/s10851-010-0251-1
+#
+#    [E+2017] M. J. Ehrhardt, P. J. Markiewicz, P. Richtarik, J. Schott,
+#    A. Chambolle and C.-B. Schoenlieb, *Faster PET reconstruction with a
+#    stochastic primal-dual hybrid gradient method*. Wavelets and Sparsity XVII,
+#    58 (2017) http://doi.org/10.1117/12.2272946.
+#
+#    [EMS2018] M. J. Ehrhardt, P. J. Markiewicz and C.-B. Schoenlieb, *Faster
+#    PET Reconstruction with Non-Smooth Priors by Randomization and
+#    Preconditioning*. (2018) ArXiv: http://arxiv.org/abs/1808.07150
+#
+# This demo is a 'script', i.e. intended to be run step by step in a
+# Python IDE such as spyder. It is organised in 'cells'. spyder displays these
+# cells nicely and allows you to run each cell on its own.
+#
+# First version: 8th of September 2018
+# Author: Matthias J Ehrhardt, Edoardo Pasca
+#
+## CCP PETMR Synergistic Image Reconstruction Framework (SIRF)
+## Copyright 2015 - 2018 Rutherford Appleton Laboratory STFC
+## Copyright 2015 - 2018 University College London.
+##
+## This is software developed for the Collaborative Computational
+## Project in Positron Emission Tomography and Magnetic Resonance imaging
+## (http://www.ccppetmr.ac.uk/).
+##
+## Licensed under the Apache License, Version 2.0 (the "License");
+##   you may not use this file except in compliance with the License.
+##   You may obtain a copy of the License at
+##       http://www.apache.org/licenses/LICENSE-2.0
+##   Unless required by applicable law or agreed to in writing, software
+##   distributed under the License is distributed on an "AS IS" BASIS,
+##   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+##   See the License for the specific language governing permissions and
+##   limitations under the License.
+import numpy
+import pSTIR as pet
+from ccpi.plugins.regularisers import FGP_TV, TGV, LLT_ROF, Diff4th
+
+class OperatorInd(object):
+
+    def __init__(self, op, subset_num, num_subsets):
+        self.__op__ = op
+        self.__subset_num__ = subset_num
+        self.__num_subsets__ = num_subsets
+
+        # FIXME: this may be known to STIR
+        x = op.img_templ.copy()
+        x.fill(1)
+        y = self.forward_sirf(x).as_array()
+        self.ind = numpy.nonzero(y.flatten())[0]
+
+    def forward_sirf(self, x):
+        return self.__op__.forward(x, subset_num=self.__subset_num__,
+                                   num_subsets=self.__num_subsets__)
+
+    def __call__(self, x):
+        y = self.__op__.forward(x, subset_num=self.__subset_num__,
+                                num_subsets=self.__num_subsets__)
+        return self.sirf2sub(y)
```

Comment:
Here is a mistake. This class assumes that "forward", "direct" and "__call__" are all the same. I think we agreed that this is not the case. Thus, 

```
y = self.__op__.forward(x, subset_num=self.__subset_num__,
                                num_subsets=self.__num_subsets__)
```

should be changed to 

```
y = self.__op__.direct(x, subset_num=self.__subset_num__,
                                num_subsets=self.__num_subsets__)
```

Reply:
Or maybe we can just remove the "__call__" altogether as both CCPi and SIRF do not support the syntax "A(x)" anyways.

Reply:
> This class assumes that "forward", "direct" and "call" are all the same. I think we agreed that this is not the case.

I believe we agreed that `forward` and `direct` were the same (after @jakobsj 's terminology post). I have directed @paskino accordingly. We introduced the `get_linear_model` for this.

I have no strong feelings about `__call__`. If we have it, it would be `forward` (but then what do we do for `adjoint` etc). Maybe it's an ODL thing?

Reply:
> I believe we agreed that forward and direct were the same (after @jakobsj 's terminology post). I have directed @paskino accordingly. We introduced the get_linear_model for this.

Ah, ok. I kind of lost track of our discussion.

> I have no strong feelings about __call__. If we have it, it would be forward (but then what do we do for adjoint etc). Maybe it's an ODL thing?

In ODL, we write `A(x)` and `A.adjoint(y)` but it might actually be better not to use it in order to avoid misunderstandings.